thoth/python/source.py

```diff
@@ -193,6 +198,109 @@ def _warehouse_get_package_hashes(
 
         return result
 
+    def _elf_find_versioned_symbols(elf: ELFFile) -> Iterator[Tuple[str, str]]:
+        section = elf.get_section_by_name(".gnu.version_r")
+
+        if section is not None:
+            for verneed, verneed_iter in section.iter_versions():
+                if verneed.name.starstwith("ld-linux"):
+                    continue
+                for vernaux in verneed_iter:
+                    yield (verneed.name, vernaux.name)
+
+    def _is_elf(filename: str):
+        with open(filename, "rb") as f:
+            return f.read(16).startswith(b'\x7f\x45\x4c\x46')
+
+    def _get_versioned_symbols_from_file(filename: str):
+        to_ret = set()
+        with open(filename, "rb") as f:
+            if not _is_elf(filename):
+                return to_ret
+            elf = ELFFile(f)
+            for _, sym in _elf_find_versioned_symbols(elf):
+                to_ret.add(sym)
+        return to_ret
+
+    def _get_versioned_symbols_in_dir(root: str):
+        to_ret = set({})
+        for dirName, subdirList, fileList in os.walk(root):
+            for fname in fileList:
+                to_ret = to_ret | _get_versioned_symbols_from_file(os.path.join(dirName, fname))
+
+        return to_ret
+
+    def _get_versioned_symbols_from_whl(whl: str):
+        tempdir = tempfile.mkdtemp()
+        try:
+            with zipfile.ZipFile(whl, "r") as zip_ref:
+                zip_ref.extractall(tempdir)
+                to_ret = _get_versioned_symbols_in_dir(tempdir)
+        except Exception as e:
+            tf = tarfile.open(whl)
+            tf.extractall(tempdir)
+            to_ret = _get_versioned_symbols_in_dir(tempdir)
+
+        shutil.rmtree(tempdir)
+        return to_ret
+
+    """
+    This function returns a tuple. 1 in position 0 represents GCC < 5,
+    1 in position 1 represents GCC > 5 these two are incompatible so it's enough
```

Comment:
I think we should be version agnostic here - there are multiple versions and more to come. We should simply get these versions in this module and store them in the knowledge graph. Then the comparison logic will go to the recommendation part.

Reply:
The reason I checked here was because of ABI conflicts between these two versions that don't exist between others.  I do agree that it should not go here though